## Singular Value Decomposition (SVD) Example using numpy (Not using scikit-learn here) ##
### Given a rectangular matrix A, it can be factored into the form $A = U \Sigma V^T$ where $U$ and $V^T$ are orthonormal and $\Sigma$ is a diagonal matrix of singular values ###


In [1]:
import numpy as np
np.set_printoptions(precision=4, suppress=True)

### We will take a simple 3x4 matrix and look at its Singular Value Decomposition with numpy ###

In [2]:
a= np.array([[1,2,3,4], [1,1,2,3],[0,1,1,0]])

### The SVD of this 3x4 matrix seeks a factorization in the form ###

$
\begin{bmatrix}
    1     & 2  & 3 & 4  \\
    1     & 1  & 2 & 3 \\
    0     & 1  & 1 & 0
\end{bmatrix}
=
\begin{bmatrix}
       &  &   \\
       & 3\times 3 ~\text{matrix}  &  \\
       &  & 
\end{bmatrix}.
\begin{bmatrix}
    \sigma_1     & 0  & 0 & 0  \\
    0     & \sigma_2  & 0 & 0  \\
    0     & 0  & \sigma_3 & 0 \\
\end{bmatrix}.
\begin{bmatrix}
       &  &   \\
       & 4\times 4 ~\text{matrix}   &  \\
       &  & 
\end{bmatrix}
 $

### The number of singular values, $\sigma$'s, is at most 3, since the original matrix has rank at most 3 ###
### The left singular matrix and the right singular matrix are orthonormal ###

In [3]:
u, s, vh = np.linalg.svd(a, full_matrices=True)

In [4]:
u.shape, s.shape, vh.shape

((3, 3), (3,), (4, 4))

In [5]:
u.shape, s.shape, vh.shape

((3, 3), (3,), (4, 4))

In [6]:
u

array([[ 0.8109, -0.0934,  0.5776],
       [ 0.57  ,  0.3493, -0.7437],
       [ 0.1323, -0.9324, -0.3365]])

In [7]:
vh

array([[ 0.2046,  0.3443,  0.5488,  0.7338],
       [ 0.2181, -0.6561, -0.438 ,  0.5746],
       [-0.7598,  0.3431, -0.4167,  0.3625],
       [-0.5774, -0.5774,  0.5774,  0.    ]])

In [8]:
s

array([6.7509, 1.1734, 0.2186])

In [9]:
np.dot(u,u.T)  # Verify u is orthonormal

array([[ 1.,  0.,  0.],
       [ 0.,  1., -0.],
       [ 0., -0.,  1.]])

In [10]:
np.dot(vh,vh.T) # Verify vh is orthonormal

array([[ 1., -0.,  0.,  0.],
       [-0.,  1.,  0., -0.],
       [ 0.,  0.,  1., -0.],
       [ 0., -0., -0.,  1.]])

### Since s is given simply as a 3-element numpy array, we need to reshape it into a 3x4 array with the 3 elements along its diagonal. Below is one way to do it - not the only way to do it. ##

In [11]:
sd=np.diag(s) # makes a 3x3 diagonal matrix with x along the diagonal

In [12]:
sd.shape

(3, 3)

In [13]:
b = np.zeros((3,4))

In [14]:
b[:,:-1]=sd

In [15]:
b

array([[6.7509, 0.    , 0.    , 0.    ],
       [0.    , 1.1734, 0.    , 0.    ],
       [0.    , 0.    , 0.2186, 0.    ]])

In [16]:
print(np.dot(np.dot(u,b),vh))  ### reconstructing a as product of u, s and vh

[[ 1.  2.  3.  4.]
 [ 1.  1.  2.  3.]
 [ 0.  1.  1. -0.]]


### So we have factored the matrix a into the form $U \Sigma V^T $ where $U$ and $V$ are orthonormal and $\Sigma$ is the diagonal matrix of singular values of the original matrix: ###

<font size=2>$
\begin{bmatrix}
    1     & 2  & 3 & 4  \\
    1     & 1  & 2 & 3 \\
    0     & 1  & 1 & 0
\end{bmatrix}=
\begin{bmatrix}
    0.8109   &-0.0934  &  0.5776 \\
     0.57   & 0.3493  & -0.7437 \\
 0.1323      &-0.9324  & -0.3365
\end{bmatrix}.
\begin{bmatrix}
      6.7509  & 0  & 0 & 0  \\
    0     & 1.1734   & 0 & 0  \\
    0     & 0  & 0.2186 & 0 \\
\end{bmatrix}.
\begin{bmatrix}
      0.2046  &   0.3443 &  0.5488 &  0.7338 \\
      0.2181 &-0.6561   & -0.438 &  0.5746\\
      -0.7598 & 0.3431  &  -0.4167 & 0.3625\\
      -0.5774&-0.5774&0.5774 &0
\end{bmatrix}
$</font>

#### Notice how the sigular values are ordered in decreasing magintue along the diagonal of $\Sigma$ ###

## Another Example - an mxn matrix where m > n ##

In [17]:
a= np.array([[1,2,3,4], [1,1,2,3],[0,1,1,0],[0,2,2,0], [0,5,5,0] ])

$ \begin{bmatrix}
1 &  2 & 3 & 4 \\
1 & 1 & 2 & 3 \\
0 & 1 & 1 & 0 \\
0 & 2 & 2 & 0 \\
0 & 5 & 5 & 0 
\end{bmatrix}  $

### We should find 3 or fewer singular values since rank of this matrix is <= 3 ###

In [18]:
u, s, vh = np.linalg.svd(a)

In [19]:
u

array([[-0.5185, -0.5983,  0.6108, -0.    , -0.    ],
       [-0.3389, -0.512 , -0.7893,  0.    ,  0.    ],
       [-0.1433,  0.1125, -0.0114,  0.0312, -0.9827],
       [-0.2866,  0.225 , -0.0229, -0.9302,  0.0383],
       [-0.7166,  0.5626, -0.0572,  0.3658,  0.1812]])

In [20]:
s

array([9.2296, 4.4454, 0.2312, 0.    ])

In [21]:
vh

array([[-0.0929, -0.6149, -0.7079, -0.3349],
       [-0.2498,  0.375 ,  0.1252, -0.8839],
       [-0.7718,  0.3846, -0.3872,  0.3264],
       [ 0.5774,  0.5774, -0.5774,  0.    ]])

In [22]:
u.shape, s.shape, vh.shape

((5, 5), (4,), (4, 4))

In [23]:
u, s, vh = np.linalg.svd(a, full_matrices=False)

In [24]:
u, s, vh

(array([[-0.5185, -0.5983,  0.6108, -0.    ],
        [-0.3389, -0.512 , -0.7893,  0.    ],
        [-0.1433,  0.1125, -0.0114,  0.0312],
        [-0.2866,  0.225 , -0.0229, -0.9302],
        [-0.7166,  0.5626, -0.0572,  0.3658]]),
 array([9.2296, 4.4454, 0.2312, 0.    ]),
 array([[-0.0929, -0.6149, -0.7079, -0.3349],
        [-0.2498,  0.375 ,  0.1252, -0.8839],
        [-0.7718,  0.3846, -0.3872,  0.3264],
        [ 0.5774,  0.5774, -0.5774,  0.    ]]))

In [25]:
b = np.zeros((5,4))

In [26]:
b[:-1,:]=np.diag(s)

In [27]:
b

array([[9.2296, 0.    , 0.    , 0.    ],
       [0.    , 4.4454, 0.    , 0.    ],
       [0.    , 0.    , 0.2312, 0.    ],
       [0.    , 0.    , 0.    , 0.    ],
       [0.    , 0.    , 0.    , 0.    ]])

In [28]:
print(np.dot(np.dot(u,b),vh))

ValueError: ignored

In [ ]:
print(a)